truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1355,6 +1365,307 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
         }
     }
 
+    @CoreMethod(names = "scrub_internal", required = 1, needsBlock = true)
+    public abstract static class ScrubNode extends YieldingCoreMethodNode {
+
+        private static final byte[] SCRUB_REPL_UTF8 = new byte[]{(byte)0xEF, (byte)0xBF, (byte)0xBD};
+        private static final byte[] SCRUB_REPL_ASCII = new byte[]{(byte)'?'};
+        private static final byte[] SCRUB_REPL_UTF16BE = new byte[]{(byte)0xFF, (byte)0xFD};
+        private static final byte[] SCRUB_REPL_UTF16LE = new byte[]{(byte)0xFD, (byte)0xFF};
+        private static final byte[] SCRUB_REPL_UTF32BE = new byte[]{(byte)0x00, (byte)0x00, (byte)0xFF, (byte)0xFD};
+        private static final byte[] SCRUB_REPL_UTF32LE = new byte[]{(byte)0xFD, (byte)0xFF, (byte)0x00, (byte)0x00};
+
+
+        @Child private IsTaintedNode isTaintedNode = IsTaintedNode.create();
+        @Child private TaintNode taintNode = TaintNode.create();
+        @Child private CallDispatchHeadNode strCompatAndValidNode = DispatchHeadNodeFactory.createMethodCall();
+        @Child private RopeNodes.MakeConcatNode makeConcatNode = RopeNodesFactory.MakeConcatNodeGen.create(null, null, null);
+
+        @Specialization
+        public DynamicObject scrubNoBlock(VirtualFrame frame, DynamicObject string, Object repl, NotProvided block) {
+            return scrubDefault(frame, string, repl, nil());
+        }
+
+        @Specialization
+        public DynamicObject scrubDefault(VirtualFrame frame, DynamicObject string, Object repl, DynamicObject block) {
+            final DynamicObject rubyEncoding = getContext().getEncodingManager().getRubyEncoding(StringOperations.encoding(string));
+            final Rope rope = rope(string);
+            final ByteList value = RopeOperations.getByteListReadOnly(rope);
+
+            CodeRange cr = rope.getCodeRange();
+            Encoding enc;
+            Encoding encidx;
+            Rope buf = null;
+            byte[] repBytes;
+            int rep;
+            int replen;
+            boolean tainted = false;
+
+            if (cr == CR_7BIT || cr == CR_VALID) {
```

Comment:
Adding a specialization for this would have the same effect.
ConditionProfile allows to compile only the branch used.

In the case the CR it 7-bit or valid we only want the fast path to get compiled.
Yes, always use createBinaryProfile(). The counting one is for advanced scenarios.

Indeed, it would be very helpful on `if (enc.isAsciiCompatible())`.